src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+
+      when (io.maskmode === 0.U) {
+        // Byte
+        readdata_mask := io.response.bits & 0xff.U
+      } .elsewhen (io.maskmode === 1.U) {
+        // Half-word
+        readdata_mask := io.response.bits & 0xffff.U
+      } .otherwise {
+        // Word
+        readdata_mask := io.response.bits
+      }
+
+      when (io.sext) {
+        when (io.maskmode === 0.U) {
+          readdata_mask_sext := Cat(Fill(24, readdata_mask(7)),  readdata_mask(7, 0))
+        } .elsewhen (io.maskmode === 1.U) {
+          readdata_mask_sext := Cat(Fill(26, readdata_mask(15)), readdata_mask(15,0))
+        } .otherwise {
+          readdata_mask_sext := readdata_mask
+        }
+      } .otherwise {
+        readdata_mask_sext := readdata_mask
+      }
+
+      io.readdata := readdata_mask_sext
+    }
+  }
+}
+
+
+/**
+ * The modified asynchronous form of the dual ported memory module. 
+ * When io.imem.request.valid or io.imem.request.valid is true and the memory is ready for an operation,
+ * this memory module simulates the latency of real DRAM by pushing memory accesses into pipes that delay 
+ * the request for a configurable latency.
+ *
+ * As with the synced memory module, this memory should only be instantiated in the Top file,
+ * and never within the actual CPU.
+ *
+ * The I/O for this module is defined in [[AsyncMemIO]].
+ */
+class DualPortedAsyncMemory(size: Int, memfile: String, latency: Int) extends Module {
+  val io = IO(new Bundle {
+    val imem = new AsyncMemIO
+    val dmem = new AsyncMemIO
+  })
+  io := DontCare
+
+  assert(latency > 0) // Check for attempt to make combinational memory
+
+  val memory    = Mem(math.ceil(size.toDouble/4).toInt, UInt(32.W))
+  loadMemoryFromFile(memory, memfile)
+
+  // Instruction port
+
+  val imemPipe = Module(new Pipe(new Request, latency))
+  val imemBusy = Reg(new Bool)
+
+  imemPipe.io := DontCare
+
+  io.imem.request.ready := ~imemBusy
+  when (io.imem.request.ready && io.imem.request.valid
+        && io.imem.request.bits.operation =/= 0.U) {
+    // Put the Request into the instruction pipe and signal that instruction memory is busy
+    val inRequest = io.imem.request.asTypeOf(new Request)
+    imemPipe.io.enq.bits <> inRequest
+    imemPipe.io.enq.valid := true.B
+    imemBusy := true.B
+  }
+ 
+  when (imemPipe.io.deq.valid && imemBusy) {
+    // Dequeue the request out of memory and execute it
+    val outRequest = imemPipe.io.deq.asTypeOf (new Request)
+    when (outRequest.operation === 1.U) { 
+      io.imem.response.valid := true.B
+      io.imem.response.bits  := memory(outRequest.address >> 2)
+    } // Ignore instruction writes as there is no way imem can issue a write access to memory
+    // Signal that instruction memory is idling as we're done
+    imemBusy := false.B
+  }
+
+  // Data port
+
+  val dmemPipe = Module(new Pipe(new Request, latency))
+  val dmemBusy = Reg(new Bool)
+
+  dmemPipe.io := DontCare
+
+  io.dmem.request.ready := ~dmemBusy
```

Comment:
You could potentially not have the duplicated code by making this a function and passing in the pipe, busy regs, and the io.